pysmt/solvers/msat.py

```diff
@@ -272,49 +354,19 @@ def get_value(self, item):
 
         titem = self.converter.convert(item)
         tval = mathsat.msat_get_model_value(self.msat_env, titem)
-
-        if mathsat.msat_term_is_number(self.msat_env, tval):
-            rep = mathsat.msat_term_repr(tval)
-            if self.environment.stc.get_type(item).is_real_type():
-                match = re.match(r"(-?\d+)/(\d+)", rep)
-                if match is not None:
-                    return self.mgr.Real((int(match.group(1)),
-                                          int(match.group(2))))
-                else:
-                    return self.mgr.Real(int(rep))
-            elif self.environment.stc.get_type(item).is_int_type():
-                return self.mgr.Int(int(rep))
-            else:
-                assert self.environment.stc.get_type(item).is_bv_type()
-                # MathSAT representation is <value>_<width>
-                value, width = rep.split("_")
-                return self.mgr.BV(int(value), int(width))
-
-        else:
-            assert mathsat.msat_term_is_true(self.msat_env, tval) or \
-                mathsat.msat_term_is_false(self.msat_env, tval)
-            bval = (mathsat.msat_term_is_true(self.msat_env, tval) == 1)
-            return self.mgr.Bool(bval)
+        val = self.converter.back(tval)
+        if self.environment.stc.get_type(item).is_real_type() and \
+               val.is_int_constant():
+            val = self.mgr.Real(val.constant_value())
+        return val
 
     def get_model(self):
-        assignment = {}
-        msat_iterator = mathsat.msat_create_model_iterator(self.msat_env)
-        while mathsat.msat_model_iterator_has_next(msat_iterator):
-            term, value = mathsat.msat_model_iterator_next(msat_iterator)
-            pysmt_term = self.converter.back(term)
-            pysmt_value = self.converter.back(value)
-            if self.environment.stc.get_type(pysmt_term).is_real_type() and \
-               pysmt_value.is_int_constant():
-                pysmt_value = self.mgr.Real(pysmt_value.constant_value())
-            assignment[pysmt_term] = pysmt_value
-        mathsat.msat_destroy_model_iterator(msat_iterator)
-        return EagerModel(assignment=assignment, environment=self.environment)
-
+        return MathSAT5Model(self.environment, self.msat_env_wrapper.ref())
 
     def exit(self):
         if not self._destroyed:
             self._destroyed = True
-            mathsat.msat_destroy_env(self.msat_env)
+            self.msat_env_wrapper.unref()
             mathsat.msat_destroy_config(self.msat_config)
```

Comment:
Do we need to guard also msat_config? From ```mathsat.h``` (msat_destroy_config):
```
 * It is an error to destroy a configuration that is still being used by an
 * environment.
```
 If so, I think we can safely assume that 1 env = 1 conf, and move the freeing of the msat_config within the ```MSatEnvWrapper```